lib/board.ex

```diff
@@ -66,28 +47,82 @@ defmodule Board do
     end
   end
 
+  def board_full?(board) do
+    board
+    |> Enum.reject(&mark?/1)
+    |> Enum.empty?
+  end
+
+  def dimension(board) do
+    length(board)
+    |> :math.sqrt
+    |> round
+  end
+
+  def current_lines(board) do
+    [rows(board), columns(board), diagonals(board)]
+    |> List.flatten
+    |> Enum.chunk(dimension(board))
+  end
+
+  def rows(board), do: Enum.chunk(board, dimension(board))
+
+  defp columns(board) do
+    board
+    |> rows
+    |> transpose
+  end
+
+  defp diagonals(board), do: diagonals(board, dimension(board))
+  defp diagonals(board, dimension) do
+    if dimension == 3 do
+      [[Enum.at(board, 0), Enum.at(board, 4), Enum.at(board, 8)],
+       [Enum.at(board, 2), Enum.at(board, 4), Enum.at(board, 6)]]
+    else
+      [[Enum.at(board, 0), Enum.at(board, 5), Enum.at(board, 10), Enum.at(board, 15)],
+       [Enum.at(board, 3), Enum.at(board, 6), Enum.at(board, 9), Enum.at(board, 12)]]
+    end
+  end
+
+  defp update_board({:valid, position}, board) do
+    next_board = board
+                  |> next_player_mark
+                  |> update(board, position - 1)
+    {:ok, next_board}
+  end
+
+  defp update_board(error, _), do: error
+
+  defp update(mark, board, position) do
+    List.replace_at(board, position , mark)
+  end
+
   defp winning_mark(_, []), do: nil
   defp winning_mark(board, [line | rest]) do
-    if winning_line(board, line) do
-      Enum.at(board, elem(line, 0))
+    if winning_line(line) do
+      List.first(line)
     else
       winning_mark(board, rest)
     end
   end
 
-  defp winner?(_, []), do: false
-  defp winner?(board, [line | rest]) do
-    if winning_line(board, line), do: true, else: winner?(board, rest)
+  defp winning_line(line) do
+    all_same_marks?(line)
   end
 
-  defp winning_line(board, {first, second, third}) do
-    same_marks?(Enum.at(board, first), Enum.at(board, second), Enum.at(board, third))
+  defp all_same_marks?(line) do
```

Comment:
This feels a little clunky. You could destructure/pattern match on [head | rest] and then just check if Enum.all?(elements, fn(element) -> element == head end)). 

If you want to stick to map and then filter, there is Enum.map_filter